BaragonData/src/main/java/com/hubspot/baragon/data/BaragonStateDatastore.java

```diff
@@ -97,27 +95,40 @@ public void removeService(String serviceId) {
   }
 
   @Timed
-  public void removeUpstreams(String serviceId, Collection<UpstreamInfo> upstreams) {
-    for (UpstreamInfo upstreamInfo : upstreams) {
-      deleteNode(String.format(UPSTREAM_FORMAT, serviceId, sanitizeNodeName(upstreamInfo.getUpstream())));
+  public void updateService(BaragonRequest request) throws Exception {
+    if (!nodeExists(SERVICES_FORMAT)) {
+      createNode(SERVICES_FORMAT);
     }
-  }
 
-  @Timed
-  public void addUpstreams(String serviceId, Collection<UpstreamInfo> upstreams) {
-    for (UpstreamInfo upstreamInfo : upstreams) {
-      writeToZk(String.format(UPSTREAM_FORMAT, serviceId, sanitizeNodeName(upstreamInfo.getUpstream())), upstreamInfo);
-    }
-  }
+    String serviceId = request.getLoadBalancerService().getServiceId();
 
-  @Timed
-  public void setUpstreams(String serviceId, Collection<UpstreamInfo> upstreams) throws Exception {
-    for (UpstreamInfo upstreamInfo : getUpstreamsMap(serviceId).values()) {
-      deleteNode(String.format(UPSTREAM_FORMAT, serviceId, sanitizeNodeName(upstreamInfo.getUpstream())));
+    CuratorTransactionFinal transaction = curatorFramework.inTransaction().create().forPath(String.format(SERVICE_FORMAT, serviceId), serialize(request.getLoadBalancerService())).and();
+
+    Map<String, byte[]> serializedAddUpstreams = new HashMap<>();
+    List<String> removeUpstreamPaths = new ArrayList<>();
+
+    if (!request.getReplaceUpstreams().isEmpty()) {
+      for (UpstreamInfo upstreamInfo : getUpstreamsMap(serviceId).values()) {
+        removeUpstreamPaths.add(String.format(UPSTREAM_FORMAT, serviceId, sanitizeNodeName(upstreamInfo.getUpstream())));
+      }
+      for (UpstreamInfo upstreamInfo : request.getReplaceUpstreams()) {
+        serializedAddUpstreams.put(String.format(UPSTREAM_FORMAT, serviceId, sanitizeNodeName(upstreamInfo.getUpstream())), serialize(upstreamInfo));
+      }
+    } else {
+      for (UpstreamInfo upstreamInfo : request.getRemoveUpstreams()) {
+        removeUpstreamPaths.add(String.format(UPSTREAM_FORMAT, serviceId, sanitizeNodeName(upstreamInfo.getUpstream())));
```

Comment:
Can you just add the delete to the transaction immediately and get rid of `removeUpstreamPaths`?